File: src/main/java/me/aov/sellgui/managers/NBTPriceManager.java
package me.aov.sellgui.managers;

import io.lumine.mythic.lib.api.item.NBTItem;
import io.lumine.mythic.lib.api.item.ItemTag;
import me.aov.sellgui.SellGUIMain;
import org.bukkit.ChatColor;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.Set;

public class NBTPriceManager {

    private final SellGUIMain main;
    private final Random random;

    private static final String PRICE_NBT_KEY = "sellgui:price";
    private static final String PRICE_TYPE_NBT_KEY = "sellgui:price_type";

    private static final String PRICE_TYPE_FIXED = "FIXED";
    private static final String PRICE_TYPE_RANDOM = "RANDOM";

    public NBTPriceManager(SellGUIMain main) {
        this.main = main;
        this.random = new Random();
    }

    public ItemStack setFixedPrice(ItemStack itemStack, double price) {
        if (itemStack == null) return itemStack;

        try {
            NBTItem nbtItem = NBTItem.get(itemStack);
            nbtItem.addTag(new ItemTag(PRICE_NBT_KEY, price));
            nbtItem.addTag(new ItemTag(PRICE_TYPE_NBT_KEY, PRICE_TYPE_FIXED));

            ItemStack result = nbtItem.toItem();
            return updateWorthLore(result, price);
        } catch (Exception e) {
            main.getLogger().warning("Failed to set fixed price: " + e.getMessage());
            return itemStack;
        }
    }

    public ItemStack setRandomPrice(ItemStack itemStack, double minPrice, double maxPrice) {
        if (itemStack == null) return itemStack;

        try {
            double finalPrice = calculateRandomPrice(minPrice, maxPrice);

            NBTItem nbtItem = NBTItem.get(itemStack);
            nbtItem.addTag(new ItemTag(PRICE_NBT_KEY, finalPrice));
            nbtItem.addTag(new ItemTag(PRICE_TYPE_NBT_KEY, PRICE_TYPE_RANDOM));
            nbtItem.addTag(new ItemTag("sellgui:min_price", minPrice));
            nbtItem.addTag(new ItemTag("sellgui:max_price", maxPrice));

            ItemStack result = nbtItem.toItem();
            return updateWorthLore(result, finalPrice);
        } catch (Exception e) {
            main.getLogger().warning("Failed to set random price: " + e.getMessage());
            return itemStack;
        }
    }

    private double calculateRandomPrice(double minPrice, double maxPrice) {
        if (minPrice >= maxPrice) {
            return minPrice;
        }
        return minPrice + (random.nextDouble() * (maxPrice - minPrice));
    }

    public double getPriceFromNBT(ItemStack itemStack) {
        if (itemStack == null) return 0.0;

        try {
            NBTItem nbtItem = NBTItem.get(itemStack);
            if (nbtItem.hasTag(PRICE_NBT_KEY)) {
                return nbtItem.getDouble(PRICE_NBT_KEY);
            }
        } catch (Exception e) {

        }

        return 0.0;
    }

    public boolean hasNBTPrice(ItemStack itemStack) {
        if (itemStack == null) return false;

        try {
            NBTItem nbtItem = NBTItem.get(itemStack);
            return nbtItem.hasTag(PRICE_NBT_KEY);
        } catch (Exception e) {
            return false;
        }
    }

    public String getPriceType(ItemStack itemStack) {
        if (itemStack == null) return null;

        try {
            NBTItem nbtItem = NBTItem.get(itemStack);
            if (nbtItem.hasTag(PRICE_TYPE_NBT_KEY)) {
                return nbtItem.getString(PRICE_TYPE_NBT_KEY);
            }
        } catch (Exception e) {

        }

        return null;
    }

    private ItemStack updateWorthLore(ItemStack itemStack, double price) {
        if (itemStack == null || !itemStack.hasItemMeta()) {
            return itemStack;
        }

        ItemMeta meta = itemStack.getItemMeta();
        List<String> lore = meta.hasLore() ? new ArrayList<>(meta.getLore()) : new ArrayList<>();

        lore.removeIf(line -> ChatColor.stripColor(line).startsWith("Worth:"));

        String worthLine = ChatColor.GOLD + "Worth: " + ChatColor.GREEN + "$" + String.format("%.2f", price);
        lore.add(worthLine);

        meta.setLore(lore);
        itemStack.setItemMeta(meta);

        return itemStack;
    }

    public ItemStack removePricing(ItemStack itemStack) {
        if (itemStack == null) return itemStack;

        try {
            NBTItem nbtItem = NBTItem.get(itemStack);
            nbtItem.removeTag(PRICE_NBT_KEY);
            nbtItem.removeTag(PRICE_TYPE_NBT_KEY);
            nbtItem.removeTag("sellgui:min_price");
            nbtItem.removeTag("sellgui:max_price");

            ItemStack result = nbtItem.toItem();

            if (result.hasItemMeta()) {
                ItemMeta meta = result.getItemMeta();
                if (meta.hasLore()) {
                    List<String> lore = new ArrayList<>(meta.getLore());
                    lore.removeIf(line -> ChatColor.stripColor(line).startsWith("Worth:"));
                    meta.setLore(lore);
                    result.setItemMeta(meta);
                }
            }

            return result;
        } catch (Exception e) {
            main.getLogger().warning("Failed to remove pricing: " + e.getMessage());
            return itemStack;
        }
    }

    public double[] getRandomPriceRange(ItemStack itemStack) {
        if (itemStack == null) return null;

        try {
            NBTItem nbtItem = NBTItem.get(itemStack);
            if (nbtItem.hasTag("sellgui:min_price") && nbtItem.hasTag("sellgui:max_price")) {
                return new double[]{
                        nbtItem.getDouble("sellgui:min_price"),
                        nbtItem.getDouble("sellgui:max_price")
                };
            }
        } catch (Exception e) {

        }

        return null;
    }

    public boolean isAvailable() {
        try {
            Class.forName("io.lumine.mythic.lib.api.item.NBTItem");
            return true;
        } catch (ClassNotFoundException e) {
            return false;
        }
    }

    public ItemStack regenerateRandomPrice(ItemStack itemStack) {
        if (itemStack == null) return itemStack;

        double[] range = getRandomPriceRange(itemStack);
        if (range != null) {
            return setRandomPrice(itemStack, range[0], range[1]);
        }

        return itemStack;
    }
}

File: src/main/kotlin/me/aov/sellgui/managers/NBTPriceManager.kt
package me.aov.sellgui.managers

import io.lumine.mythic.lib.api.item.ItemTag
import io.lumine.mythic.lib.api.item.NBTItem
import me.aov.sellgui.SellGUIMain
import org.bukkit.ChatColor
import org.bukkit.inventory.ItemStack

class NBTPriceManager(private val main: SellGUIMain) {
    private val random = java.util.Random()

    fun setFixedPrice(itemStack: ItemStack?, price: Double): ItemStack? {
        if (itemStack == null) return null
        return try {
            val nbtItem = NBTItem.get(itemStack)
            nbtItem.addTag(ItemTag(PRICE_NBT_KEY, price))
            nbtItem.addTag(ItemTag(PRICE_TYPE_NBT_KEY, PRICE_TYPE_FIXED))
            updateWorthLore(nbtItem.toItem(), price)
        } catch (exception: Exception) {
            main.logger.warning("Failed to set fixed price: ${exception.message}")
            itemStack
        }
    }

    fun setRandomPrice(itemStack: ItemStack?, minPrice: Double, maxPrice: Double): ItemStack? {
        if (itemStack == null) return null
        return try {
            val finalPrice = calculateRandomPrice(minPrice, maxPrice)
            val nbtItem = NBTItem.get(itemStack)
            nbtItem.addTag(ItemTag(PRICE_NBT_KEY, finalPrice))
            nbtItem.addTag(ItemTag(PRICE_TYPE_NBT_KEY, PRICE_TYPE_RANDOM))
            nbtItem.addTag(ItemTag(MIN_PRICE_NBT_KEY, minPrice))
            nbtItem.addTag(ItemTag(MAX_PRICE_NBT_KEY, maxPrice))
            updateWorthLore(nbtItem.toItem(), finalPrice)
        } catch (exception: Exception) {
            main.logger.warning("Failed to set random price: ${exception.message}")
            itemStack
        }
    }

    private fun calculateRandomPrice(minPrice: Double, maxPrice: Double): Double =
        if (minPrice >= maxPrice) minPrice else minPrice + random.nextDouble() * (maxPrice - minPrice)

    fun getPriceFromNBT(itemStack: ItemStack?): Double = try {
        if (itemStack != null && NBTItem.get(itemStack).hasTag(PRICE_NBT_KEY)) NBTItem.get(itemStack).getDouble(PRICE_NBT_KEY) else 0.0
    } catch (_: Exception) {
        0.0
    }

    fun getSellPrice(itemStack: ItemStack?): Double = getPriceFromNBT(itemStack)

    fun setSellPrice(itemStack: ItemStack?, price: Double) {
        setFixedPrice(itemStack, price)
    }

    fun hasNBTPrice(itemStack: ItemStack?): Boolean = try {
        itemStack != null && NBTItem.get(itemStack).hasTag(PRICE_NBT_KEY)
    } catch (_: Exception) {
        false
    }

    fun getPriceType(itemStack: ItemStack?): String? = try {
        itemStack?.let { NBTItem.get(it) }?.takeIf { it.hasTag(PRICE_TYPE_NBT_KEY) }?.getString(PRICE_TYPE_NBT_KEY)
    } catch (_: Exception) {
        null
    }

    private fun updateWorthLore(itemStack: ItemStack, price: Double): ItemStack {
        if (!itemStack.hasItemMeta()) return itemStack
        val meta = itemStack.itemMeta
        val lore = if (meta.hasLore()) ArrayList(meta.lore) else ArrayList()
        lore.removeIf { ChatColor.stripColor(it)?.startsWith("Worth:") == true }
        lore.add("${ChatColor.GOLD}Worth: ${ChatColor.GREEN}$${"%.2f".format(price)}")
        meta.lore = lore
        itemStack.itemMeta = meta
        return itemStack
    }

    fun removePricing(itemStack: ItemStack?): ItemStack? {
        if (itemStack == null) return null
        return try {
            val nbtItem = NBTItem.get(itemStack)
            listOf(PRICE_NBT_KEY, PRICE_TYPE_NBT_KEY, MIN_PRICE_NBT_KEY, MAX_PRICE_NBT_KEY).forEach(nbtItem::removeTag)
            nbtItem.toItem().also { result ->
                if (result.hasItemMeta()) {
                    val meta = result.itemMeta
                    if (meta.hasLore()) {
                        meta.lore = ArrayList(meta.lore).also { it.removeIf { line -> ChatColor.stripColor(line)?.startsWith("Worth:") == true } }
                        result.itemMeta = meta
                    }
                }
            }
        } catch (exception: Exception) {
            main.logger.warning("Failed to remove pricing: ${exception.message}")
            itemStack
        }
    }

    fun getRandomPriceRange(itemStack: ItemStack?): DoubleArray? = try {
        val nbtItem = itemStack?.let(NBTItem::get) ?: return null
        if (nbtItem.hasTag(MIN_PRICE_NBT_KEY) && nbtItem.hasTag(MAX_PRICE_NBT_KEY)) doubleArrayOf(nbtItem.getDouble(MIN_PRICE_NBT_KEY), nbtItem.getDouble(MAX_PRICE_NBT_KEY)) else null
    } catch (_: Exception) {
        null
    }

    fun isAvailable(): Boolean = try {
        Class.forName("io.lumine.mythic.lib.api.item.NBTItem")
        true
    } catch (_: ClassNotFoundException) {
        false
    }

    fun regenerateRandomPrice(itemStack: ItemStack?): ItemStack? {
        val range = getRandomPriceRange(itemStack)
        return if (range != null) setRandomPrice(itemStack, range[0], range[1]) else itemStack
    }

    private companion object {
        const val PRICE_NBT_KEY = "sellgui:price"
        const val PRICE_TYPE_NBT_KEY = "sellgui:price_type"
        const val PRICE_TYPE_FIXED = "FIXED"
        const val PRICE_TYPE_RANDOM = "RANDOM"
        const val MIN_PRICE_NBT_KEY = "sellgui:min_price"
        const val MAX_PRICE_NBT_KEY = "sellgui:max_price"
    }
}
